exercises/forth/test/Tests.hs

```diff
@@ -17,62 +17,82 @@ specs = describe "forth" $ do
-      runTexts ["1 2 3 over"] `shouldBe` Right [1, 2, 3, 2]
-      runTexts ["1 over"    ] `shouldBe` Left StackUnderflow
-      runTexts ["over"      ] `shouldBe` Left StackUnderflow
-
-    it "defining a new word" $
-      runTexts [ ": dup-twice dup dup ;"
-               , "1 dup-twice"           ] `shouldBe` Right [1, 1, 1]
-
-    it "redefining an existing word" $
-      runTexts [ ": foo dup ;"
-               , ": foo dup dup ;"
-               , "1 foo"           ] `shouldBe` Right [1, 1, 1]
-
-    it "redefining an existing built-in word" $
-      runTexts [ ": swap dup ;"
-               , "1 swap"       ] `shouldBe` Right [1, 1]
-
-    it "defining words with odd characters" $
-      runTexts [": € 220371 ; €"] `shouldBe` Right [220371]
-
-    it "defining a number" $
-      runTexts [": 1 2 ;"] `shouldBe` Left InvalidWord
-
-    it "calling a non-existing word" $
-      runTexts ["1 foo"] `shouldBe` Left (UnknownWord "foo")
+    describe "parsing and numbers" $ do
+      it "empty input results in empty stack" $
+        toList empty `shouldBe` []
+
+      it "numbers just get pushed onto the stack" $
+        runTexts ["1 2 3 4 5"] `shouldBe` Right [1, 2, 3, 4, 5]
+
+      it "all non-word characters are separators" $
+        runTexts ["1\NUL2\SOH3\n4\r5 6\t7"] `shouldBe` Right [1, 2, 3, 4, 5, 6, 7]
+
+    describe "division" $
+      it "errors if dividing by zero" $
+        runTexts ["4 0 /"] `shouldBe` Left DivisionByZero
+
+    describe "combined arithmetic" $ do
+      it "addition and subtraction" $
+        runTexts ["1 2 + 4 -"] `shouldBe` Right [-1]
+
+      it "multiplication and division" $
+        runTexts ["2 4 * 3 /"] `shouldBe` Right [2]
+
+    describe "dup" $ do
+      it "copies the top value on the stack" $
+        runTexts ["1 DUP"  ] `shouldBe` Right [1, 1]
+      it "is case-insensitive" $
+        runTexts ["1 2 Dup"] `shouldBe` Right [1, 2, 2]
+      it "errors if there is nothing on the stack" $
+        runTexts ["dup"    ] `shouldBe` Left StackUnderflow
+
+    describe "drop" $ do
+      it "removes the top value on the stack if it is the only one" $
+        runTexts ["1 drop"  ] `shouldBe` Right []
+      it "removes the top value on the stack if it is not the only one" $
+        runTexts ["1 2 drop"] `shouldBe` Right [1]
+      it "errors if there is nothing on the stack" $
+        runTexts ["drop"    ] `shouldBe` Left StackUnderflow
+
+    describe "swap" $ do
+      it "swaps the top two values on the stack if they are the only ones" $
+        runTexts ["1 2 swap"  ] `shouldBe` Right [2, 1]
+      it "swaps the top two values on the stack if they are not the only ones" $
+        runTexts ["1 2 3 swap"] `shouldBe` Right [1, 3, 2]
+      it "errors if there is nothing on the stack" $
+        runTexts ["swap"      ] `shouldBe` Left StackUnderflow
+      it "errors if there is only one value on the stack" $
+        runTexts ["1 swap"    ] `shouldBe` Left StackUnderflow
+
```

Comment:
Oh, I was inconsistent. `1 over` is before `over`. I'll make it consistent, and note an order change in commit message.